File: planner/src/receive.ts
#!/usr/bin/env node

import amqp from 'amqplib'
import dotenv from 'dotenv'
import { PrismaDataSource } from './datasources/Prisma'
import { BasicPlanner } from './planners/BasicPlanner'
import { Planner } from './planners/Planner'

dotenv.config()

const datasource = new PrismaDataSource()
const planner: Planner = new BasicPlanner(datasource)

async function main() {
  console.log('url is', process.env.AMQP_URL)
  const connection = await amqp.connect(
    process.env.AMQP_URL || 'amqp://localhost'
  )

  const channel = await connection.createChannel()

  const queue = process.env.QUEUE_NAME || 'planner'

  await channel.assertQueue(queue, { durable: false })

  console.log(' [*] Waiting for messages in %s. To exit press CTRL+C', queue)

  await channel.consume(
    queue,
    async (msg) => {
      console.log(' [x] Received %s', msg?.content.toString())
      if (!msg?.content) {
        console.log(' [x] No content')
        return
      }
      await onMessageReceived(msg.content.toString())
      console.log(' [x] Finished')
    },
    { noAck: true }
  )
}

async function onMessageReceived(msg: string) {
  try {
    const message = JSON.parse(msg) as { planId: string }
    if (!message.planId || typeof message.planId !== 'string') {
      console.log(' [x] No planId in message: %s', msg)
      return
    }
    const plans = await planner.start(message.planId)
    if (!plans.success) {
      console.log(' [x] Failed to plan %s', message.planId)
      return
    }
    console.log(' [x] Planned %d jobs', plans.jobs.length)
    datasource.setPlannedJobs(message.planId, plans.jobs)
  } catch (e) {
    console.log(' [x] Failed to parse message: %s', msg)
    console.error(e)
  }
}

main()

File: planner/src/receive.py
#!/usr/bin/env python
from planner.src.solver import generate_plan_from_message
from rabbitmq_setup import setup_connection, QUEUE_NAME

channel = setup_connection()

def on_message(ch, method, body):
    print(f'Received message (delivery tag: {method.delivery_tag}): {body}')
    generate_plan_from_message("9fdcbb17-5ade-4a68-a51d-1a9e7dc9e10b")
    ch.basic_ack(delivery_tag=method.delivery_tag)

channel.basic_consume(queue=QUEUE_NAME, on_message_callback=on_message, auto_ack=False)

print(' [*] Waiting for messages. To exit press CTRL+C')
channel.start_consuming()